### PR TITLE
[RFC][WIP] Extend convert_to_unspecialized for module attr mutation to module fields mutated through BINARY_SUBSCR

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106163

fixes

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel @anijain2305

### CODE REVIEW

**Comment by voznesenskym** [comment]
https://github.com/pytorch/pytorch/actions/runs/5685516023

**Comment by voznesenskym** [comment]
The simple version of PR absolutely reks perf. Looks like I cant take the easy way out.

File: test/inductor/test_inductor_freezing.py [hunk 1]
```diff
@@ -217,6 +217,10 @@ def foo(mod, inp):
                 ).run(code[0])
                 self.assertEqual(out_eager, out)
 
+    # freezing_discard_parameters is an unsound optimization, and it does not work with the attribute
+    # lifting strategy in dynamo. Dynamo needs this strategy to preserve correctness, which trumps speed
```
Comment (eellison) [inline test/inductor/test_inductor_freezing.py:221]: it's not an unsound optimization, and we should find a better strategy than just disabling existing tests

Comment (voznesenskym) [reply test/inductor/test_inductor_freezing.py]: fine

**Comment by voznesenskym** [comment]
perf neutral

File: torch/_dynamo/symbolic_convert.py [hunk 2]
```diff
@@ -1251,6 +1252,11 @@ def should_compile_partial_graph(self) -> bool:
     @break_graph_if_unsupported(push=0)
     def STORE_SUBSCR(self, inst):
         val, obj, key = self.popn(3)
+        if obj.source and obj.source.guard_source().is_nn_module():
+            module_source = get_nn_module_source(obj.source)
+            module = module_source.find_module(self)
+            NNModuleVariable.convert_module_to_unspecialized(self, module)
```
Comment (ezyang) [inline torch/_dynamo/symbolic_convert.py:1258]: This cannot possibly be complete. What about all of the other ways you can mutate a list?

Comment (voznesenskym) [reply torch/_dynamo/symbolic_convert.py:1258]: This isn't even close to complete. I am just exploring, hence the WIP. 

I am working on a refactor that identifies all "mutating" operations and routes them through a central fn where we can make decisions like this.

Comment (ezyang) [reply torch/_dynamo/symbolic_convert.py:1258]: Don't we already have this, isn't it side effects?

**Pull request closed without merging**